### ISSUE TITLE
swagger annotations break linter

### ISSUE THREAD

**Issue by lucasvo** [uber/prototool#69]
I'm trying to get the linter to pass on my swagger annotations however it seems like protoc is invoked differently for `prototool lint` command than for `prototool gen`. The modifications below to the example break lint but work for gen.  Is this an issue with my configuration or are the swagger annotations not supported yet?

```bash,
# prototool lint
[...]
2018-05-08T12:07:26.706-0400	DEBUG	running protoc	{"command": "/Users/lucasvo/Library/Caches/prototool/Darwin/x86_64/protobuf/3.5.1/bin/protoc -I /Users/lucasvo/Code/go/src/github.com/uber/prototool/vendor/github.com/grpc-ecosystem/grpc-gateway -I /Users/lucasvo/Code/go/src/github.com/uber/prototool/vendor/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis -I /Users/lucasvo/Library/Caches/prototool/Darwin/x86_64/protobuf/3.5.1/include -I /Users/lucasvo/Code/go/src/github.com/uber/prototool/example/idl/uber -o /dev/null /Users/lucasvo/Code/go/src/github.com/uber/prototool/example/idl/uber/foo/foo.proto"}
foo/foo.proto:18:1: found "}" but expected [.proto element {comment|option|import|syntax|enum|service|package|message}]

# prototool gen
[...]
2018-05-08T12:07:05.189-0400	DEBUG	running protoc	{"command": "/Users/lucasvo/Library/Caches/prototool/Darwin/x86_64/protobuf/3.5.1/bin/protoc -I /Users/lucasvo/Code/go/src/github.com/uber/prototool/vendor/github.com/grpc-ecosystem/grpc-gateway -I /Users/lucasvo/Code/go/src/github.com/uber/prototool/vendor/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis -I /Users/lucasvo/Library/Caches/prototool/Darwin/x86_64/protobuf/3.5.1/include -I /Users/lucasvo/Code/go/src/github.com/uber/prototool/example/idl/uber --gogoslick_out=plugins=grpc,Mgoogle/protobuf/compiler/plugin.proto=github.com/gogo/protobuf/protoc-gen-gogo/plugin,Mgoogle/protobuf/descriptor.proto=github.com/gogo/protobuf/protoc-gen-gogo/descriptor,Mgoogle/protobuf/duration.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/source_context.proto=google.golang.org/genproto/protobuf/source_context,Mgoogle/protobuf/type.proto=google.golang.org/genproto/protobuf/ptype,Mgoogle/protobuf/any.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/api.proto=google.golang.org/genproto/protobuf/api,Mgoogle/protobuf/empty.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/field_mask.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/struct.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/timestamp.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/wrappers.proto=github.com/gogo/protobuf/types,Mgoogle/api/http.proto=google.golang.org/genproto/googleapis/api/annotations,Mprotoc-gen-swagger/options/annotations.proto=github.com/grpc-ecosystem/grpc-gateway/protoc-gen-swagger/options/annotations.proto,Mgoogle/api/annotations.proto=google.golang.org/genproto/googleapis/api/annotations:/Users/lucasvo/Code/go/src/github.com/uber/prototool/example/gen/proto/go /Users/lucasvo/Code/go/src/github.com/uber/prototool/example/idl/uber/foo/foo.proto"}
2018-05-08T12:07:05.189-0400	DEBUG	running protoc	{"command": "/Users/lucasvo/Library/Caches/prototool/Darwin/x86_64/protobuf/3.5.1/bin/protoc -I /Users/lucasvo/Code/go/src/github.com/uber/prototool/vendor/github.com/grpc-ecosystem/grpc-gateway -I /Users/lucasvo/Code/go/src/github.com/uber/prototool/vendor/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis -I /Users/lucasvo/Library/Caches/prototool/Darwin/x86_64/protobuf/3.5.1/include -I /Users/lucasvo/Code/go/src/github.com/uber/prototool/example/idl/uber --java_out=/Users/lucasvo/Code/go/src/github.com/uber/prototool/example/gen/proto/java /Users/lucasvo/Code/go/src/github.com/uber/prototool/example/idl/uber/foo/foo.proto"}
```

The diff for the example:
```diff,
diff --git a/example/idl/uber/foo/foo.proto b/example/idl/uber/foo/foo.proto
index 7e94743..0d8e47a 100644
--- a/example/idl/uber/foo/foo.proto
+++ b/example/idl/uber/foo/foo.proto
@@ -5,6 +5,7 @@ import "google/protobuf/timestamp.proto";
 
 import "foo/dep.proto";
 import "google/api/annotations.proto";
+import "protoc-gen-swagger/options/annotations.proto";
 import "sub/sub.proto";
 
 package foo;
@@ -12,6 +13,9 @@ package foo;
 option go_package = "foopb";
 option java_multiple_files = true;
 option java_package = "com.foo.pb";
+option (grpc.gateway.protoc_gen_swagger.options.openapiv2_swagger) = {
+  host: "example.com"; 
+};
 
 // Hello is a hello.
 enum Hello {
diff --git a/example/idl/uber/prototool.yaml b/example/idl/uber/prototool.yaml
index 592b48e..5177db7 100644
--- a/example/idl/uber/prototool.yaml
+++ b/example/idl/uber/prototool.yaml
@@ -2,6 +2,7 @@
 protoc_includes:
   # note vendor is .gitignored
   - ../../../vendor/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis
+  - ../../../vendor/github.com/grpc-ecosystem/grpc-gateway
 protoc_include_wkt: true
 lint:
   # run "prototool list-linters" to see the currently configured linters
@@ -17,6 +18,8 @@ gen:
     extra_modifiers:
       google/api/annotations.proto: google.golang.org/genproto/googleapis/api/annotations
       google/api/http.proto: google.golang.org/genproto/googleapis/api/annotations
+      protoc-gen-swagger/options/annotations.proto: github.com/grpc-ecosystem/grpc-gateway/protoc-gen-swagger/options/annotations.proto
+
   plugins:
     - name: gogoslick
       type: gogo
```

**Comment by peter-edge** [comment]
The linter runs through protoc first, and then also runs through emicklei/proto's parser, and this is a bug in emicklei/proto that is already filed: https://github.com/emicklei/proto/issues/79

Going to close this one as this is effectively a duplicate of that bug, is that OK?

**Comment by lucasvo** [comment]
It seems like it's fixed in emicklei/proto. Perhaps we can update the version of the library. I'm trying that out right now. Also, the thing that sent me on a wild goose chase is that the debug log made me believe that the validation was coming straight out of protoc because the preceding message printed the protoc command invocation. Perhaps it would be a good idea to add an explanation to where the message is coming from.

**Comment by peter-edge** [comment]
Can you link to the point where you’d want additional log messaging? I’m happy to add it. You can put up a PR too if you wish :-) but it’s quick for me.

I’m waiting on another bug fix before updating the emicklei/proto import again, but if the update you do fixes it, I can update immediately.

**Comment by peter-edge** [comment]
Good idea. Will put up a PR tomorrow that adds this logging.

**Comment by johanbrandhorst** [comment]
This works if you remove the semicolon at the end of the line:
```
option (grpc.gateway.protoc_gen_swagger.options.openapiv2_swagger) = {
    host: "example.com";
};
```

**Comment by lucasvo** [comment]
You are right, the following syntax works:

```
option (grpc.gateway.protoc_gen_swagger.options.openapiv2_swagger) = {
    host: "localhost:8082"
    info: {
        title:"Sample API"
    }
    schemes:HTTPS
};
```

Which is very different than what grpc-gateway uses in it's sample: 

https://github.com/grpc-ecosystem/grpc-gateway/blob/master/examples/proto/examplepb/a_bit_of_everything.proto#L13

I don't know the protobuf spec well enough, is the extra semicolon allowed? It seems like this should be a lint error only and not a syntax error if it is an accepted and commonly used format.

**Comment by peter-edge** [comment]
If it compiles with `protoc`, it should work in `emickei/proto` - I'm forgetting where we left off here, is there an active bug for this filed on `emicklei/proto`?

**Comment by johanbrandhorst** [comment]
I think https://github.com/emicklei/proto/issues/79 was closed incorrectly. This issue is that issue.

**Comment by lucasvo** [comment]
It seems like the linter is not the only thing that is broken. `prototool format` turns valid protobuf syntax into something that isn't accepted by `prototool gen`:

```
option (grpc.gateway.protoc_gen_swagger.options.openapiv2_swagger) = {
  host: "localhost:8082"
  info {
      title: "Test API"
      version: "0.0.1"
  }
  schemes: HTTPS
  consumes: "application/json"
  produces: "application/json"
};
```
Running prototool all produces the following output:
```
MBLAV:centrifuge-protobufs lucasvo$ prototool all
invoice/service.proto:12:70:1:1:Error while parsing option value for "openapiv2_swagger": Expected "{", found ".".
invoice/service.proto:12:70:1:1:Error while parsing option value for "openapiv2_swagger": Expected "{", found ".".
```

And the above swagger definition gets changed to: 
```
option (grpc.gateway.protoc_gen_swagger.options.openapiv2_swagger) = {
  host: "localhost:8082"
  info.title: "Test API"
  info.version: "0.0.1"
  schemes: HTTPS
  consumes: "application/json"
  produces: "application/json"
};
```

**Comment by peter-edge** [comment]
This is fixed on dev, this was https://github.com/uber/prototool/issues/27.

**Comment by peter-edge** [comment]
To install dev, do make install.

**Comment by peter-edge** [comment]
Can someone see if this is still broken on dev? This may be fixed now that everything is updated.

**Comment by lucasvo** [comment]
Yes, both issues still seem to persist. I just tested against 1f691826e56012e947fd37cb1f739ef54ba575b1. You can apply the patch I added in the issue description to test.

**Comment by peter-edge** [comment]
The commit that should have fixed it was after that and was today - I can try the patch, but if you pull and `make install` and have it readily available, maybe easier verification?

**Comment by lucasvo** [comment]
Still broken.

**Comment by johanbrandhorst** [comment]
This is fixed for me on `dev`.

**Comment by peter-edge** [comment]
Oh yea - you have to run glide install again after pulling.

**Comment by peter-edge** [comment]
I've verified both of the issues here (the semicolon issue and the format issue) are fixed on dev. When you `git pull` dev, make sure you run `make init` again to get the latest vendored updates.

If this is still not fixed for someone, re-open, but closing for now.

**Comment by johanbrandhorst** [comment]
I'm glad this is fixed in dev - but there being no release with this fix still prohibits me from using prototool at work. Any chance we could cut a release for this fix?

**Comment by peter-edge** [comment]
Release just cut. Sorry for the delay.
